JavaScript/JavaScript.sublime-syntax

```diff
@@ -1088,59 +1088,76 @@ contexts:
+
+  literal-double-quoted-string:
+    - match: \"
       scope: punctuation.definition.string.begin.js
-      set:
-        - meta_include_prototype: false
-        - meta_scope: meta.string.js string.quoted.single.js
-        - match: \'
-          scope: punctuation.definition.string.end.js
-          pop: true
-        - match: \n
-          scope: invalid.illegal.newline.js
-          pop: true
-        - include: string-content
-    - match: '"'
+      set: literal-double-quoted-string-content
+
+  literal-double-quoted-string-content:
+    - meta_include_prototype: false
+    - meta_scope: meta.string.js string.quoted.double.js
+    - match: \"
+      scope: punctuation.definition.string.end.js
+      pop: true
+    - match: \n
+      scope: invalid.illegal.newline.js
+      pop: true
+    - include: string-content
+
+  literal-single-quoted-string:
+    - match: \'
       scope: punctuation.definition.string.begin.js
-      set:
-        - meta_include_prototype: false
-        - meta_scope: meta.string.js string.quoted.double.js
-        - match: \"
-          scope: punctuation.definition.string.end.js
-          pop: true
-        - match: \n
-          scope: invalid.illegal.newline.js
-          pop: true
-        - include: string-content
+      set: literal-single-quoted-string-content
+
+  literal-single-quoted-string-content:
+    - meta_include_prototype: false
+    - meta_scope: meta.string.js string.quoted.single.js
+    - match: \'
+      scope: punctuation.definition.string.end.js
+      pop: true
+    - match: \n
+      scope: invalid.illegal.newline.js
+      pop: true
+    - include: string-content
 
   literal-string-template:
-    - match: '`'
+    - match: \`
       scope: punctuation.definition.string.begin.js
-      set:
-        - meta_include_prototype: false
-        - meta_scope: meta.string.js string.quoted.other.js
-        - match: "`"
-          scope: punctuation.definition.string.end.js
-          pop: true
-        - match: '\$\{'
-          scope: punctuation.section.interpolation.begin.js
-          push:
-            - clear_scopes: 1
-            - meta_scope: meta.interpolation.js
-            - meta_content_scope: source.js.embedded
-            - match: '\}'
-              scope: punctuation.section.interpolation.end.js
-              pop: true
-            - match: (?=\S)
-              push: expression
-        - include: string-content
+      set: literal-string-template-content
+
+  literal-string-template-content:
+    - meta_include_prototype: false
+    - meta_scope: meta.string.js string.quoted.other.js
+    - match: \`
+      scope: punctuation.definition.string.end.js
+      pop: true
+    - include: string-interpolations
+    - include: string-content
 
   string-content:
     - match: \\\n
       scope: constant.character.escape.newline.js
     - match: \\(?:x\h\h|u\h\h\h\h|.)
       scope: constant.character.escape.js
 
+  string-interpolations:
+    - match: \$\{
```

Comment:
Does it make sense to give the leading `$` a different scope like `punctuation.definition.interpolation.js` since it signals the start of an interpolation ?

Reply:
This kind of `$` interpolation scoping is a topic on its own, I guess. I struggled with scoping them in Bash/Perl/PHP and probably various other syntaxes when working on variables and interpolation topics. Especially interesting in syntaxes like PHP/Perl/Bash whose variables all begin with `$`. So it is normally scoped `punctuation.definition.variable`, while being scoped `punctuation.section.interpolation` in some contexts (like here).

Perl also uses it as `referencing operator` in case multiple `$$$$var` appear.

We should probably revice that with all related syntaxes being looked into.

Reply:
My main point here was scoping of `$` within strings that are considered interpolations. `$` in variables receiving `punctuation.definition.variable` makes sense, but I don't know anything that uses it apart from PHP and perhaps that's a related but separate topic than strings. But I agree, it can be tackled separately.